include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,113 @@
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   */
+  virtual void sendMessage(const RequestType& request) PURE;
+
+  /**
+   * Close the stream locally. No further methods may be invoked on the
+   * stream object, but callbacks may still be received until the stream is closed remotely.
+   */
+  virtual void close() PURE;
+
+  /**
+   * Close the stream locally and remotely (as needed). No further methods may be invoked on the
+   * stream object and no further callbacks will be invoked.
+   */
+  virtual void reset() PURE;
+};
+
+/**
+ * Notifies caller of async gRPC stream status.
+ * Note the gRPC stream is full-duplex, even if the local to remote stream has been ended by
+ * AsyncClientStream.close(), AsyncClientCallbacks can continue to receive events until the remote
+ * to local stream is closed (onRemoteClose), and vice versa. Once the stream is closed remotely, no
+ * further callbacks will be invoked.
+ */
+template <class ResponseType> class AsyncClientCallbacks {
+public:
+  virtual ~AsyncClientCallbacks() {}
+
+  /**
+   * Called when populating the headers to send with initial metadata.
+   * @param metadata initial metadata reference.
+   */
+  virtual void onCreateInitialMetadata(Http::HeaderMap& metadata) PURE;
+
+  /**
+   * Called when initial metadata is recevied.
+   * @param metadata initial metadata reference.
+   */
+  virtual void onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) PURE;
+
+  /**
+   * Called when an async gRPC message is received.
+   * @param response the gRPC message.
+   */
+  virtual void onReceiveMessage(std::unique_ptr<ResponseType>&& message) PURE;
+
+  /**
+   * Called when trailing metadata is recevied.
+   * @param metadata trailing metadata reference.
+   */
+  virtual void onReceiveTrailingMetadata(Http::HeaderMapPtr&& metadata) PURE;
+
+  /**
+   * Called when the remote closes or an error occurs on the gRPC stream. The stream is
+   * considered remotely closed after this invocation and no further callbacks will be
+   * invoked. A non-Ok status implies that stream is also locally closed and that no
+   * further stream operations are permitted.
+   * @param status the gRPC status.
+   */
+  virtual void onRemoteClose(Status::GrpcStatus status) PURE;
+};
+
+/**
+ * Supports sending gRPC requests and receiving responses asynchronously. This can be used to
+ * implement either plain gRPC or streaming gRPC calls.
+ */
+template <class RequestType, class ResponseType> class AsyncClient {
+public:
+  virtual ~AsyncClient() {}
+
+  /**
+   * Start a gRPC stream asynchronously.
+   * @param service_method protobuf descriptor of gRPC service method.
+   * @param callbacks the callbacks to be notified of stream status.
+   * @param timeout supplies the stream timeout, measured since when the frame with end_stream
```

Comment:
We had [previous discussion](https://github.com/lyft/envoy/pull/353#discussion_r96561929) around this. The intended behavior for `Http::AsyncClient` is what you have comment here, since it is also used for unary. I think this only make sense for unary calls, so if you want use the AsyncClient for unary call too, probably just leave this here and comment it so.